src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/ProgramDescriptor.java

```diff
@@ -0,0 +1,119 @@
+ * or deletion without notice.</b></p>
+ */
+public record ProgramDescriptor(ProgramFileObject fileObject, Optional<String> packageName, Path sourceRootPath) {
+    static ProgramDescriptor of(ProgramFileObject fileObject) throws Fault {
+        var file = fileObject.getFile();
+        try {
+            var compiler = JavacTool.create();
+            var standardFileManager = compiler.getStandardFileManager(null, null, null);
+            var units = List.of(fileObject);
+            var task = compiler.getTask(null, standardFileManager, diagnostic -> {}, null, null, units);
+            for (var tree : task.parse()) {
+                var packageTree = tree.getPackage();
+                if (packageTree != null) {
+                    var packageName = packageTree.getPackageName().toString();
+                    var root = computeSourceRootPath(file, packageName);
+                    return new ProgramDescriptor(fileObject, Optional.of(packageName), root);
+                }
+            }
+        } catch (IOException ignore) {
+            // fall through to let actual compilation determine the error message
+        }
+        var root = computeSourceRootPath(file, "");
+        return new ProgramDescriptor(fileObject, Optional.empty(), root);
+    }
+
+    public static Path computeSourceRootPath(Path program, String packageName) {
+        var absolute = program.normalize().toAbsolutePath();
+        var absoluteRoot = absolute.getRoot();
+        assert absoluteRoot != null;
+        // unnamed package "": program's directory is the root path
+        if (packageName.isEmpty()) {
+            var parent = absolute.getParent();
+            if (parent == null) return absoluteRoot;
+            return parent;
+        }
+        // named package "a.b.c": ensure end of path to program is "a/b/c"
+        var packagePath = Path.of(packageName.replace('.', '/'));
+        var ending = packagePath.resolve(program.getFileName());
+        if (absolute.endsWith(ending)) {
+            var max = absolute.getNameCount() - ending.getNameCount();
+            if (max == 0) return absoluteRoot;
+            return absoluteRoot.resolve(absolute.subpath(0, max));
+        }
+        throw new Fault(Errors.MismatchEndOfPathAndPackageName(packageName, program));
+    }
+
+    public boolean isModular() {
+        return Files.exists(sourceRootPath.resolve("module-info.java"));
+    }
+
+    public Set<String> computePackageNames() {
+        try (var stream = Files.find(sourceRootPath, 99, (path, attr) -> attr.isDirectory())) {
+            var names = new TreeSet<String>();
+            stream.filter(ProgramDescriptor::containsAtLeastOneRegularFile)
+                  .map(sourceRootPath::relativize)
+                  .map(Path::toString)
+                  .filter(string -> !string.isEmpty())
+                  .map(string -> string.replace('/', '.'))
+                  .map(string -> string.replace('\\', '.'))
```

Comment:
Nit: I'd be a bit careful with the `\` -> `.` replacement. `\` is a valid file name character on UNIX-like system. Better use `File.separatorChar` or `sourceRootPath.getFileSystem().getSeparator()`. Although the practical impact is probably very low.

Reply:
Addressed this in https://github.com/openjdk/jdk/pull/13712/commits/093ab1cef7383323deb74bffa511b81109ca4dd6